### PR TITLE
Feature persistmsg

### PR DESCRIPTION
**Pull request by urviljoshi**

### CODE REVIEW

File: src/repository/msgpersistence.bal [hunk 1]
```diff
@@ -81,6 +81,32 @@ public type MessagePersistenceImpl object {
         return messageDetails;
     }
 
+    public function findMessageByID(string ID) returns @tainted MessageDetails{
```
Comment (vishwa-vyom) [inline src/repository/msgpersistence.bal:84]: As discussed lets do a list of messageId and handle with in query

File: src/services/hubservice.bal [hunk 2]
```diff
@@ -104,6 +106,38 @@ public type HubServiceImpl object {
...
     }
 
+
+    public function getFailedContent(string subscriberSignature,string topic , string callback ,string timestamp ,int messagecount) returns @untainted repository:FailedContentPullRespModel[]|error{
+               repository:SubscriptionExtendedDetails subscriptionExtendedDetails = self.subsOperations.getSubscription(topic,callback);
+               string hmacSubsSignature=utils:hmacSha256(topic+callback+timestamp,subscriptionExtendedDetails.secret);
+               if(hmacSubsSignature!=subscriberSignature){
+                   return error("SIGNATUREMATCHERROR", message = "hmac didnot match");
+               }
+               string[] msgIDs=self.deliveryReportPersistence.getFailedDeliveryBySubID(subscriptionExtendedDetails.id,timestamp,messagecount);
```
Comment (vishwa-vyom) [inline src/services/hubservice.bal:116]: As discussed lets do a update and select in a single SQL query and use that to optimize the DB calls

**Review by vishwa-vyom — COMMENTED** [review]
Please check the comments

**Comment by urviljoshi** [comment]
@vishwa-vyom  can you look at the review changes.

File: src/hubinit/constants.bal [hunk 3]
```diff
@@ -0,0 +1,10 @@
+const string TOPIC_EMPTY_ERROR_MESSAGE= "topic should be a non empty string ";
+const string CALLBACK_EMPTY_ERROR_MESSAGE= "callback should be a non empty string ";
+const string TIMESTAMP_EMPTY_ERROR_MESSAGE= "timestamp should be a non empty string ";
+const string TIMESTAMP_FORMAT_ERROR_MESSAGE= "timestamp should be in ISO format ";
+const string SUBSCRIBER_SIGNATURE_HEADER = "X-Subscriber-Signature";
+const string TOPIC = "topic";
+const string CALLBACK = "callback";
+const string TIMESTAMP = "timestamp";
+const string MESSAGECOUNT = "messageCount";
+const string TIMESTAMP_PATTERN = "yyyy-MM-dd'T'HH:mm:ss.SSS";
```
[the remark this answers is not in the document]
Comment (vishwa-vyom) [reply src/hubinit/constants.bal]: This is available in the latest code

File: src/hubinit/failedcontentpull.bal [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+             return ();
+        }
+        if(timestamp is string && timestamp!= ""){
+         timestampParameter=<string>timestamp;
+         }else{
+            check caller->badRequest(TIMESTAMP_EMPTY_ERROR_MESSAGE);
+             return ();
+        }
+        if(messageCountValue is int){
+         messageCountParameter= messageCountValue;
```
[the remark this answers is not in the document]
Comment (vishwa-vyom) [reply src/hubinit/failedcontentpull.bal:53]: This is available in the latest code

File: src/repository/msgpersistence.bal (hunk 1 above)
Comment (vishwa-vyom) [reply src/repository/msgpersistence.bal]: This is available in the latest code

File: src/services/hubservice.bal (hunk 2 above)
Comment (vishwa-vyom) [reply src/services/hubservice.bal]: This is available in the latest code

File: src/hubinit/failedcontentpull.bal [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            check caller->badRequest(CALLBACK_EMPTY_ERROR_MESSAGE);
+             return ();
+        }
+        if(timestamp is string && timestamp!= ""){
+         timestampParameter=<string>timestamp;
+         }else{
+            check caller->badRequest(TIMESTAMP_EMPTY_ERROR_MESSAGE);
+             return ();
+        }
+        if(messageCountValue is int){
```
[the remark this answers is not in the document]
Comment (vishwa-vyom) [reply src/hubinit/failedcontentpull.bal:48]: I see this is not taken care, @urviljoshi can you take care of this?

**Review by vishwa-vyom — APPROVED** [review]
Hi Urvil,

I see the below review point alone is not taken care, please fix it in subsequent commits, for now I am approving the changes.

'Can we also have a max allowed count validation based on a config value?'

**Review by vishwa-vyom — APPROVED** [review]
Please go ahead and merge

**Pull request merged**